tests/suite/advanced_routing_test.go

```diff
@@ -0,0 +1,282 @@
+			httpRoute := "http-header-and-query-matching"
+			grpcRoute := "grpc-header-matching"
+
+			nsname := types.NamespacedName{Name: httpRoute, Namespace: namespace}
+			err := waitForHTTPRouteToBeAccepted(nsname)
+			Expect(err).ToNot(HaveOccurred(), fmt.Sprintf("%s was not accepted", httpRoute))
+
+			nsname = types.NamespacedName{Name: grpcRoute, Namespace: namespace}
+			err = waitForGRPCRouteToBeAccepted(nsname)
+			Expect(err).ToNot(HaveOccurred(), fmt.Sprintf("%s was not accepted", httpRoute))
+		})
+
+		DescribeTable("verify working traffic for HTTPRoute",
+			func(uri string, serverName string, headers map[string]string, queryParams map[string]string) {
+				url := baseURL + uri
+				Eventually(
+					func() error {
+						return expectRequestToRespondFromExpectedServer(url, address, serverName, headers, queryParams)
+					}).
+					WithTimeout(timeoutConfig.ContainerRestartTimeout).
+					WithPolling(500 * time.Millisecond).
+					Should(Succeed())
+			},
+			Entry("request with no headers or params", "/coffee", "coffee-v1", nil, nil),
+			Entry("request with Exact match header", "/coffee", "coffee-v2", map[string]string{"version": "v2"}, nil),
+			Entry("request with Exact match query param", "/coffee", "coffee-v2", nil, map[string]string{"TEST": "v2"}),
+			Entry(
+				"request with RegularExpression match header",
+				"/coffee",
+				"coffee-v3",
+				map[string]string{"headerRegex": "header-regex"},
+				nil,
+			),
+			Entry(
+				"request with RegularExpression match query param",
+				"/coffee",
+				"coffee-v3",
+				nil,
+				map[string]string{"queryRegex": "query-regex"},
+			),
+			Entry(
+				"request with non-matching regex header",
+				"/coffee",
+				"coffee-v1",
+				map[string]string{"headerRegex": "headerInvalid"},
+				nil,
+			),
+			Entry(
+				"request with non-matching regex query param",
+				"/coffee",
+				"coffee-v1",
+				nil,
+				map[string]string{"queryRegex": "queryInvalid"},
+			),
+		)
+	})
+})
+
+func expectRequestToRespondFromExpectedServer(
+	appURL, address, expServerName string,
+	headers, queryParams map[string]string,
+) error {
+	status, body, err := framework.Get(appURL, address, timeoutConfig.RequestTimeout, headers, queryParams)
+	if err != nil {
+		return err
+	}
+
+	if status != http.StatusOK {
+		return errors.New("http status was not 200")
+	}
+
+	actualServerName, err := extractServerName(body)
+	if err != nil {
+		return err
+	}
+
+	if !strings.Contains(actualServerName, expServerName) {
+		return errors.New("expected response body to contain correct server name")
+	}
+
+	return err
+}
+
+func extractServerName(responseBody string) (string, error) {
+	re := regexp.MustCompile(`Server name:\s*(\S+)`)
+	matches := re.FindStringSubmatch(responseBody)
+	if len(matches) < 2 {
+		return "", errors.New("server name not found")
+	}
+	return matches[1], nil
+}
+
+func waitForHTTPRouteToBeAccepted(nsname types.NamespacedName) error {
```

Comment:
Can we not just reuse the existing `waitForHTTPRoutesToBeReady` function that we have? Same for grpc